helix-core/src/main/java/org/apache/helix/util/InstanceUtil.java

```diff
@@ -0,0 +1,176 @@
+package org.apache.helix.util;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.ImmutableSet;
+import org.apache.helix.AccessOption;
+import org.apache.helix.BaseDataAccessor;
+import org.apache.helix.ConfigAccessor;
+import org.apache.helix.HelixException;
+import org.apache.helix.PropertyPathBuilder;
+import org.apache.helix.constants.InstanceConstants;
+import org.apache.helix.manager.zk.ZKHelixDataAccessor;
+import org.apache.helix.manager.zk.ZkBaseDataAccessor;
+import org.apache.helix.model.ClusterTopologyConfig;
+import org.apache.helix.model.HelixConfigScope;
+import org.apache.helix.model.InstanceConfig;
+import org.apache.helix.model.builder.HelixConfigScopeBuilder;
+import org.apache.helix.zookeeper.api.client.RealmAwareZkClient;
+import org.apache.helix.zookeeper.datamodel.ZNRecord;
+import org.apache.helix.zookeeper.zkclient.DataUpdater;
+
+public class InstanceUtil {
+  private final BaseDataAccessor<ZNRecord> _baseDataAccessor;
+  private final ConfigAccessor _configAccessor;
+
+  public InstanceUtil(RealmAwareZkClient zkClient) {
+    _baseDataAccessor = new ZkBaseDataAccessor<>(zkClient);
+    _configAccessor = new ConfigAccessor(zkClient);
+  }
+
+  public InstanceUtil(ConfigAccessor configAccessor, BaseDataAccessor<ZNRecord> baseAccessor) {
+    _baseDataAccessor = baseAccessor;
+    _configAccessor = configAccessor;
+  }
+
+  public static void validateInstanceOperationTransition(InstanceConfig matchingLogicalIdInstance,
```

Comment:
Why not build a static map like:

ENABLE -> Set(DISABLE, EVACUTATE) 
DISABLE. -> Set(....)

So the logic will just be 

map.get(currentOperation).contains(futureOperation).

Reply:
Turned it into a map for easier readability. Each targetOperation has a validator corresponding to its entry in the map.